### PR TITLE
Adds a new admin-spawn antagonist, The Slasher.

### PR DESCRIPTION
**Pull request by Zonespace27**
[FEATURE]
## About the PR 
NOTE: This was made over the course of a few months, there is most likely some Bad Code here, trying to iron it out where I can)

This PR adds a new admin-spawn antagonist, the Slasher.
The Slasher is an antagonist inspired by, as you can guess, slasher movies. As such, their kit is highly focused around murder.
They have high stats, making them a force to be reckoned with, as seen below:
100% Cold Resist
100% Fire Resist
95% Explosion Resist
10% Disease Resist
75% Stun Resist
75% Radiation Resist
7 Melee Body Resist
6 Melee Head Resist
0.6 Movement Delay
3 Ranged Resist
100% Eye Disorient Resist
Their abilities go as follows: (See them in video [HERE](https://www.youtube.com/watch?v=dt6PmtMtmuM))

- Incorporealize: The Slasher is capable of going incorporeal, provided there's nobody that can see the Slasher. Going incorporeal makes them: Unable to interact with the world, able to clip through walls, unable to die, and makes them invisible.
- Corporealize: The Slasher is capable of going corporeal, which reverses the effects of going incorporeal. Unlike going incorporeal, the Slasher is capable of doing it in vision of people, but a short delay + darkness effect can tip off people, potentially letting them get away.
- Soul Steal: The Slasher is capable of stealing the soul of a dead body, which raises the damage of their Machete by 2.5 damage.
- Blood Trail: This ability doesn't have much of a purpose, it just lets you trail blood infinitely (until you toggle it again) for the spookiness factor.
- Summon Machete: This ability will summon a "Slasher's Machete". If the Slasher already has one in existence, it will be called to their hand. The machete does 15 damage, has the "wide swing" special attack, stuns + deals 75% of the melee damage on throw, and cannot be picked up or dragged by crewmembers.
- Possession: [STILL A WIP] This ability is an odd one. The Slasher is able to target a person, who will, over the next 30 seconds, get scary messages and a chance of falling unconscious, potentially warning everyone that the person will be possessed soon. Once the thirty seconds pass, the target will fall unconscious, and then a "slasher's gas mask", "slasher's machete", and a "slasher's apron" will be placed on the target, with anything in those slots being dropped. Then, the target is forced into an observer focused on their own body, with the Slasher's client taking over the target. For the next 45 seconds, The Slasher's client will have full control over the target, allowing them to attack the rest of the crew, without exposing themselves. However, there are some caveats. The possessed person is much weaker, health wise, than the Slasher, the Slasher's client doesn't have any abilities for the duration of possession, in addition to the Machete not scaling up with damage (being a flat 20), and being unable to be thrown. Once the 45 seconds end, the Slasher's client is sent back to their body (The Slasher's body itself), and the target regains control of their body. The machete dissolves into ash, and any equipment the Slasher's possession gave the target is transmuted into a weaker, mostly decorative version.
- Regenerate: The Slasher gets the `full_heal()` proc executed on them, meaning that stamina is fully regenned, limbs are regrown, handcuffs are removed, and health is reset, among other things.
- Weakening Field: Everyone in a small radius gets 8 seconds of weakened, slowing them down quite a bit.

TODO:
- [x] Fix possession to work properly, the target gets stuck observing their body once the Slasher leaves it, after the 45 seconds end.
- [x] Add edge-case fixes to possession (what if the target leaves, what if the welder leaves, what if the target gets gibbed, etc.)

## Why's this needed? 
[This card](https://github.com/goonstation/goonstation/projects/2#card-38790682) wanted a rework of the Welder, so I thought I'd give it a try in time for Spooktober.

## Changelog
```changelog
(u)Zonespace
(*)There's a new admin-spawn antagonist on the loose, try not to get stabbed!
```

### CODE REVIEW

**Comment by Emcee-Gore** [comment]
Cool stuff! 100%

**Comment by colossusqw** [comment]
This looks very cool, but to be honest, i'm not much of a fan of admin spawn only antags, do you think this could work with a rare spawn like wraith is or is it just too much to play out unchecked?

**Comment by Zonespace27** [comment]
> This looks very cool, but to be honest, i'm not much of a fan of admin spawn only antags, do you think this could work with a rare spawn like wraith on RP is or is it just too much to play out unchecked?

As it stands the Slasher is powerful enough to roll the station with ease, I'm debating making it appear on friday the 13th days, assuming I make the non-admin spawn one a bit less powerful to bring it a bit more in line, and also getting dev signoff on that.

**Comment by Emcee-Gore** [comment]
> 
> 
> > This looks very cool, but to be honest, i'm not much of a fan of admin spawn only antags, do you think this could work with a rare spawn like wraith on RP is or is it just too much to play out unchecked?
> 
> As it stands the Slasher is powerful enough to roll the station with ease, I'm debating making it appear on friday the 13th days, assuming I make the non-admin spawn one a bit less powerful to bring it a bit more in line, and also getting dev signoff on that.

Spooktober and friday the 13th normal spawn antag? That could be dope.

**Comment by Emcee-Gore** [comment]
Given the tanky nature of the slasher; being able to completely heal up as well as having an ability to hinder movement of others sends some strong Halloween movie vibes.

I wonder how it might play if you slow down the base-movement speed of the slasher; ramp up the antags stamina for a longer sprint - and crank up that movement-debuff ability in strength?

**A hard to kill; slow moving menace.**

**Comment by Zonespace27** [comment]
> Given the tanky nature of the slasher; being able to completely heal up as well as having an ability to hinder movement of others sends some strong Halloween movie vibes.
> 
> 
> 
> I wonder how it might play if you slow down the base-movement speed of the slasher; ramp up the antags stamina for a longer sprint - and crank up that movement-debuff ability in strength?
> 
> 
> 
> **A hard to kill; slow moving menace.**

Forgot to mention this in the PR- the Slasher's unremovable equipment gives a cumulative 0.6 movement slowdown, equivalent to a space suit, from memory

**Comment by Zonespace27** [comment]
Covered most possession edge cases, tested them as well as I could, marking it as ready to review

**Comment by Zonespace27** [comment]
Overall dev opinion is that they aren't looking for new player-accessible antags currently, leaving it as admin-spawn unless a dev wants to chime in.

File: code/mob/dead/target_observer.dm [hunk 1]
```diff
@@ -152,3 +152,32 @@ var/list/observers = list()
...
 
+/mob/dead/target_observer/slasher_ghost
+	name = "spooky not-quite ghost"
+	var/start_time
+
+	New()
+		..()
+		APPLY_MOB_PROPERTY(src, PROP_INVISIBILITY, src, INVIS_GHOST)
+		observers += src
+		mobs += src
```
Comment (pali6) [inline code/mob/dead/target_observer.dm:163]: Don't these three already happen in `..()`?

File: config/admins.txt [hunk 2]
```diff
@@ -3,3 +3,4 @@
 #Most of the time, you'll want to make yourself a Host.
 #To do so, type your BYOND key with no spaces, capitals or other punctuation (including underscores!), a dash with spaces on both side, then the rank you want.
 #For example, the key "ZeWaka" would be given Host rank like this: zewaka - Host
+zonespace - Host
```
Comment (pali6) [inline config/admins.txt:6]: Pls remove.

Comment (Zonespace27) [reply config/admins.txt:6]: Forgot to remove this when i tested it on dream daemon with a second account, sorry!

File: code/mob/living/carbon/human/slasher.dm [hunk 3]
```diff
@@ -0,0 +1,581 @@
...
+		src.see_invisible = 16
+		src.bioHolder.AddEffect("breathless", 0, 0, 0, 1)
+		src.bioHolder.AddEffect("food_rad_resist", 0, 0, 0, 1)
+		src.bioHolder.AddEffect("detox", 0, 0, 0, 1)
+		src.add_stun_resist_mod("slasher_stun_resistance", 75)
+		START_TRACKING
+
+	Life()
+		var/turf/T = get_turf(src)
+		if(!src.density && T && isrestrictedz(T.z))
```
[the remark this answers is not in the document]
Comment (pali6) [reply code/mob/living/carbon/human/slasher.dm]: In the meantime I remembered that one source of removing density is laying down. This code will make slasher gib if they lay down in an adventure zone which seems... not ideal.

[the remark this answers is not in the document]
Comment (Zonespace27) [reply code/mob/living/carbon/human/slasher.dm]: fuck, i thought i got rid of all of the density checks

[the remark this answers is not in the document]
Comment (Zonespace27) [reply code/mob/living/carbon/human/slasher.dm]: Actually, i already did, this code chunk is outdated, see line 30 of `slasher.dm`

File: code/mob/living/carbon/human/slasher.dm [hunk 4]
```diff
@@ -0,0 +1,562 @@
...
+		send_machete_to_target(var/obj/item/I)
+			if(!istype(I))
+				return
+
+			I.visible_message("<span class='alert'><b>The [I.name] is suddenly warped away!</b></span>")
+			elecflash(I)
+
+			if(ismob(src.loc))
+				src.u_equip(I)
+			if(istype(locate(I), /obj/item/storage))
```
Comment (pali6) [inline code/mob/living/carbon/human/slasher.dm:173]: Haven't we talked about how `locate(object)` always returns null already?

Comment (Zonespace27) [reply code/mob/living/carbon/human/slasher.dm:173]: Bit of context:
this is older code from when i got using length() over .len confused with using locate() over .loc (which are two different things completely, as you know)

**Review by pali6 — APPROVED** [review]
I don't have the energy to do another full review. I think the code is better than it used to be and certainly much better than The Welder. There are things that I don't like but they'd take a lot of work to change and it's likely just not worth it.

I'm gonna merge the PR as it's Halloween tomorrow and there could be some fun rounds with this antagonist if an admin decides to use it. However, if some major issues pop up I reserve the right to revert the merge (but of course you can keep working on improving the antagonist whether it gets reverted or not).

**Comment by Zonespace27** [comment]
> I don't have the energy to do another full review. I think the code is better than it used to be and certainly much better than The Welder. There are things that I don't like but they'd take a lot of work to change and it's likely just not worth it.
> 
> I'm gonna merge the PR as it's Halloween tomorrow and there could be some fun rounds with this antagonist if an admin decides to use it. However, if some major issues pop up I reserve the right to revert the merge (but of course you can keep working on improving the antagonist whether it gets reverted or not).

Thank you Pali! I will keep this in mind!

**Pull request merged**